### PR TITLE
Definable distance pagination for ScrollView

### PR DESCRIPTION
**Pull request by rxb**
This is an enhancement for ScrollView that adds the ability to paginate based on a width other than the width of the ScrollView itself. This is a fairly common pattern used on apps like Facebook, App Store, and Twitter to scroll through a horizontal set of cards or icons:

img_8726 2 img_8727 2  img_8728 2

After trying to accomplish this only with JS, it appears that attempting to take over an in-progress native scroll animation with JS is always going to result in some amount of jankiness and jumping. 

This pull request uses `scrollViewWillEndDragging` in RCTScrollView.m to adjust `targetContentOffset` based on two new optional props added to ScrollView. `snapToInterval` sets the multiple that the ScrollView will come to rest on. `snapToAlignment` sets the relative alignment of the snap `start`, `center`, or `end`.  

Here's an example of it in action:
https://vid.me/aoby

Here is some sample code to try it out yourself:

```
var SnapTest = React.createClass({

  render: function() {
    var blocks = [];
    for(var i=0; i<=30; i++){
      blocks.push("i");
    }
    var blockW = 100;
    return (
      <ScrollView 
        horizontal={true} 
        decelerationRate={0}
        snapToInterval={blockW}
        snapToAlignment="start"
        >
        {blocks.map(function(m,i){
          return (
            <View key={i} style={{width: blockW, height: blockW, padding:16, backgroundColor: (i%2==0)?'cornflowerblue':'royalblue', borderWidth: 1, borderColor: "#fff"}}>
              <Text style={{color: 'white', fontWeight: '800'}}>{i}</Text>
            </View>
          );
        })}
      </ScrollView>
    );
  }
});
```

Would love to get your thoughts on this direction! 

cc: @brentvatne https://github.com/facebook/react-native/issues/1362

### CODE REVIEW

**Comment by brentvatne** [comment]
This looks great, I'll read over more carefully and try it out tonight :smile:

File: React/Views/RCTScrollView.m [hunk 1]
```diff
@@ -560,6 +560,54 @@ - (void)scrollViewWillBeginDragging:(UIScrollView *)scrollView
...
+  if(self.snapToInterval){
+    
+    CGFloat snapToIntervalF = (CGFloat)self.snapToInterval;
+
+    // Find which axis to snap
+    BOOL isHorizontal = (scrollView.contentSize.width > self.frame.size.width);
+
+    // What is the current offset?
+    CGFloat targetContentOffsetAlongAxis = isHorizontal ? targetContentOffset->x : targetContentOffset->y;
+    //CGFloat targetContentOffsetAlongAxis = isHorizontal ? scrollView.contentOffset.x : scrollView.contentOffset.y;
```
Comment (brentvatne) [inline React/Views/RCTScrollView.m]: This comment can go

File: React/Views/RCTScrollView.m [hunk 2]
```diff
@@ -560,6 +560,54 @@ - (void)scrollViewWillBeginDragging:(UIScrollView *)scrollView
...
+  // An alternative to enablePaging which allows setting custom stopping intervals,
+  // smaller than a full page size. Often seen in apps which feature horizonally
+  // scrolling items. snapToInterval does not enforce scrolling one interval at a time
+  // but guarantees that the scroll will stop at an interval point.
+  if(self.snapToInterval){
+    
+    CGFloat snapToIntervalF = (CGFloat)self.snapToInterval;
+
+    // Find which axis to snap
+    BOOL isHorizontal = (scrollView.contentSize.width > self.frame.size.width);
```
Comment (brentvatne) [inline React/Views/RCTScrollView.m]: Is this a safe way to determine whether or not it is horizontal? Could it ever occur that our primary scroll axis is vertical but the contentSize is both greater than the width and height of the scrollView?

File: React/Views/RCTScrollView.m [hunk 3]
```diff
@@ -560,6 +560,54 @@ - (void)scrollViewWillBeginDragging:(UIScrollView *)scrollView
...
+    CGPoint translation = [scrollView.panGestureRecognizer translationInView:scrollView];
+    CGFloat translationAlongAxis = isHorizontal ? translation.x : translation.y;
+    
+    // Offset based on desired alignment 
+    CGFloat frameLength = isHorizontal ? self.frame.size.width : self.frame.size.height;
+    CGFloat alignmentOffset = 0.0f;
+    if([self.snapToAlignment  isEqualToString: @"center"]){
+      alignmentOffset = (frameLength * 0.5f) + (snapToIntervalF * 0.5f);
+    }
+    else if ([self.snapToAlignment  isEqualToString: @"end"]) {
```
Comment (brentvatne) [inline React/Views/RCTScrollView.m]: nit: this should be up on the same line as `}`

File: React/Views/RCTScrollView.m [hunk 4]
```diff
@@ -560,6 +560,54 @@ - (void)scrollViewWillBeginDragging:(UIScrollView *)scrollView
...
+    if(translationAlongAxis < 0) {
+      snapIndex++;
+    }
+    CGFloat newTargetContentOffset = ( snapIndex * snapToIntervalF ) - alignmentOffset;
+    
+    // Set new targetContentOffset
+    if(isHorizontal) {
+      targetContentOffset->x = newTargetContentOffset;
+    }
+    else {
```
Comment (brentvatne) [inline React/Views/RCTScrollView.m]: nit: this should be up on the same line as `}`

File: React/Views/RCTScrollView.m [hunk 5]
```diff
@@ -560,6 +560,54 @@ - (void)scrollViewWillBeginDragging:(UIScrollView *)scrollView
...
+    if([self.snapToAlignment  isEqualToString: @"center"]){
+      alignmentOffset = (frameLength * 0.5f) + (snapToIntervalF * 0.5f);
+    }
+    else if ([self.snapToAlignment  isEqualToString: @"end"]) {
+      alignmentOffset = frameLength;
+    }
+
+    // Pick snap point based on direction and proximity
+    NSInteger snapIndex = floor((targetContentOffsetAlongAxis + alignmentOffset) / snapToIntervalF);
+    if(translationAlongAxis < 0) {
```
Comment (brentvatne) [inline React/Views/RCTScrollView.m]: this is more would be more concise with a ternary expression

**Comment by brentvatne** [comment]
@rxb - could you also squash these commits into one?

**Comment by brentvatne** [comment]
Not sure who owns ScrollView cc @tadeuzagallo @nicklockwood

File: React/Views/RCTScrollView.m (hunk 2 above)
Comment (rxb) [inline React/Views/RCTScrollView.m]: Was a little unsure about that one... I borrowed it from [line 411]( https://github.com/facebook/react-native/blob/41859e594d0dbd0be43cf7cc1561051ac4b1b853/React/Views/RCTScrollView.m#L411 )
```
- (void)setStickyHeaderIndices:(NSIndexSet *)headerIndices
{
  RCTAssert(_scrollView.contentSize.width <= self.frame.size.width,
           @"sticky headers are not supported with horizontal scrolled views");
  _scrollView.stickyHeaderIndices = headerIndices;
}
```

File: React/Views/RCTScrollView.m (hunk 3 above)
Comment (rxb) [inline React/Views/RCTScrollView.m]: fixed

File: React/Views/RCTScrollView.m (hunk 4 above)
Comment (rxb) [inline React/Views/RCTScrollView.m]: fixed

File: React/Views/RCTScrollView.m (hunk 1 above)
Comment (rxb) [inline React/Views/RCTScrollView.m]: fixed

File: React/Views/RCTScrollView.m (hunk 5 above)
Comment (rxb) [inline React/Views/RCTScrollView.m]: fixed

**Comment by rxb** [comment]
@brentvatne Thanks for the notes. Everything except for the [isHorizontal question](https://github.com/facebook/react-native/pull/1532#discussion_r31882481) is fixed and squashed into 1 commit.

**Comment by rxb** [comment]
What do you think about tweaking the api to match up with the CSS scroll-snap-points spec?
http://www.w3.org/TR/css-snappoints-1/#scroll-snap-type 

Looks like it will be in most browsers soon:
http://caniuse.com/#feat=css-snappoints 
https://www.chromestatus.com/features/5721832506261504

**Comment by brentvatne** [comment]
@rxb - Absolutely! I think the closer that we can be to the w3 spec the better. Great point, can you adjust accordingly? cc @vjeux

**Comment by vjeux** [comment]
Have you tried paginate={true} on ScrollView, doesn't it solve your use case?

**Comment by brentvatne** [comment]
@vjeux - there are several problems with `paginate={true}`:

- Fixed page size to the viewport width
  - If it were to support variable widths, it would need to have an option to specify where to snap to (middle, left-side, right-side)
- Horizontal paging only

**Comment by ide** [comment]
You can support variable widths with pagination with a couple of tricks -- that is how the Apple Photos app works.

The core difference between pagination and scroll snapping is that pagination snaps to the next page even if you swipe hard, while scroll snapping accounts for momentum so you could end up several pages away.

**Comment by rxb** [comment]
I think where that difference matters most is when you are scrolling through items that are not close to the screen width, like photos on FB profiles:
https://vid.me/V4Bz

It would be rough if it forced one-at-a-time, but free scrolling would feel like driving on a greased track.

File: React/Views/RCTScrollView.m (hunk 2 above)
Comment (sahrens) [inline React/Views/RCTScrollView.m]: I think this is fairly safe, although things might get weird in a 2-axis scrolling situation...the JS component has a `horizontal` prop - maybe we should pass that through to native?

**Comment by sahrens** [comment]
@nicklockwood: what do you think of this?

**Comment by rxb** [comment]
@brentvatne Awesome, I'll sync things up with the W3C spec.

After testing this in a more realistic screen -- where the horizontal scrolling section is one part of a longer vertical scrollview -- ran into that nested scrolling bug from: https://github.com/facebook/react-native/issues/41 .

**Comment by vjeux** [comment]
Just wanted to said that I patched this to prototype a use case that we wanted to do internally and it works really well :)

Sorry it takes so long to review :(

File: React/Views/RCTScrollView.m [hunk 6]
```diff
@@ -593,6 +593,51 @@ - (void)scrollViewWillBeginDragging:(UIScrollView *)scrollView
...
+    
+    // Set new targetContentOffset
+    if(isHorizontal) {
+      targetContentOffset->x = newTargetContentOffset;
+    } else {
+      targetContentOffset->y = newTargetContentOffset;
+    }
+
+  }
+<<<<<<< HEAD
```
Comment (paramaggarwal) [inline React/Views/RCTScrollView.m]: merge conflict artifact?

Comment (rxb) [inline React/Views/RCTScrollView.m]: Thanks for the catch, fixed.

**Comment by rxb** [comment]
@vjeux Happy to hear it worked for you. 

I keep meaning to get back to this and tweak to match the W3C scroll snap points spec. Would love to get anyone's thoughts on this interface:

**Enable snapping**
http://www.w3.org/TR/css-snappoints-1/#scroll-snap-type
**prop:** `scroll-snap-type` 
**values:** `none` `proximity` (`mandatory` could be supported in the future )

**Set snapping interval** 
http://www.w3.org/TR/css-snappoints-1/#propdef-scroll-snap-points-x
**prop:** `scroll-snap-points` (W3C spec defines `scroll-snap-points-x` and `scroll-snap-points-y` independently, but it seems that RN encourages scrolling in one direction at a time)
**values:** `none` `repeat([interval in pixels])` (The `repeat` seems unnecessary, but it is part of the spec.)

**Set alignment of snap relative to scroll container** 
http://www.w3.org/TR/css-snappoints-1/#propdef-scroll-snap-destination
**prop:** `scroll-snap-destination`
**values:** `[offset in pixels]` (Spec defines values for x and y offsets together, but this seems unnecessary if RN ScrollViews are only scrolling in one direction. Also setting this explicitly in pixels gives more flexibility, but seems work than `start`, `center`, `end` )

**Comment by sahrens** [comment]
ping @nicklockwood

**Comment by chirag04** [comment]
+1

Just stumbled upon this while looking to detect page changes on a horizontal listview. This could be useful i guess.

**Comment by chirag04** [comment]
@vjeux Does it make sense to emulate css api if it is not part of style? 

My 2cents: I liked the initial api suggested by @rxb.

**Comment by vjeux** [comment]
The guideline is to first try to design the best API based on what exists on iOS, Android, web... Then once we found one, if it is close enough to the web one, it's preferable to use it. If the web one doesn't make sense, we shouldn't tie ourself to it

**Comment by chirag04** [comment]
any thoughts one this one?

cc @nicklockwood

**Comment by chirag04** [comment]
@vjeux  Mind sharing the patch? Wondering if anyone is having any opinions on this. Happy to implement thoughts that others may have.

**Comment by syrusakbary** [comment]
Would love having this patch merged into react native!
:+1:

**Comment by rxb** [comment]
If there's anything I can do to help get this moving, let me know! I've been using in my personal project and it's been pretty useful.

**Comment by chirag04** [comment]
@rxb Def want to see this landed and want to use it in our project as well. It is still compatible with RN 0.11?

@vjeux  @nicklockwood  Can we land this one?

**Comment by vjeux** [comment]
@facebook-github-bot shipit

**Comment by chirag04** [comment]
Awesome. Thanks :+1:

**Comment by vjeux** [comment]
"Failed to apply patch". @rxb would you be willing to rebase? Sorry for the wait

**Comment by rxb** [comment]
Should be back in business, rebased with master. The Travis e2e test is failing, but I can't tell if that's because of something in this PR or a general thing that's happening. Seems like almost all the recent PRs are failing that.

**Comment by vjeux** [comment]
Travis is broken because of our android push. Unlikely your fault

**Comment by vjeux** [comment]
@facebook-github-bot shipit

**Comment by rxb** [comment]
Thanks, everybody!

**Comment by vjeux** [comment]
Thanks for sticking with us and sorry it took that much time!

**Comment by aleclarson** [comment]
I'm surprised an `onSnap` prop wasn't included. Thoughts?

**Comment by erickreutz** [comment]
Any plans to bring this functionality to Android?

**Comment by me-abhinav** [comment]
+1 for Android

**Comment by me-abhinav** [comment]
I can create a pull request if the owner wants.

**Pull request closed without merging**